content/docs/cucumber/state.md

```diff
@@ -238,6 +238,211 @@ compile group: 'io.cucumber', name: 'cucumber-needle', version: '{{% version "cu
 
 There is no documentation yet, but the code is on [GitHub](https://github.com/cucumber/cucumber-jvm/tree/master/needle).
 
+# Details of using DI (e.g. using Guice)
+
+(since Cucumber 5)
+
+It is typically very simple to use Cucumber with a dependency injection framework.
+All that's needed is to add the appropriate dependency to the classpath.
+
+For example, if you want to use [Google Guice](https://github.com/google/guice) as DI
+simply add the `cucumber-guice` dependency to your pom.xml:
+
+```xml
+<dependencies>
+  [...]
+    <dependency>
+        <groupId>io.cucumber</groupId>
+        <artifactId>cucumber-guice</artifactId>
+        <version>${cucumber.version}</version>
+        <scope>test</scope>
+    </dependency>
+  [...]
+</dependencies>
+```
+
+Now all your step definitions, hooks, transformers, etc. will be created and supplied by a Guice injector.
+This is pretty cool, but depending on your application it is far off from being sufficient.
+
+## The need for a custom injector
+
+Even though example tests are very simple, they often do not stay that simple when it comes
+to large applications contexts. An application often has specific components (data managers, providers,
+services, etc.). These components need to be made available to your step definitions so that actions
+can be applied on them and delivered results can be tested.
+
+The reason using Cucumber with Guice DI typically originates from the fact that the tested application also uses
+Guice as DI framework. So why not bringing this together and provide a custom injector that can inject
+the application components into the step definitions.
+
+```java
+package com.example.app;
+
+import static org.junit.Assert.assertTrue;
+
+import io.cucumber.java.en.When;
+import io.cucumber.java.en.Then;
+import io.cucumber.guice.ScenarioScoped;
+import com.example.app.service.AppService;
+import java.util.Objects;
+import javax.inject.Inject;
+
+@ScenarioScoped
+public final class StepDefinition {
+
+    private final AppService appService;
+
+    @Inject
+    public StepDefinition( AppService appService ) {
+        this.appService = Objects.requireNonNull( appService, "appService must not be null" );
+    }
+
+    @When("the application services are started")
+    public void startServices() {
+        this.appService.startServices();
+    }
+
+    @Then("all application services should be running")
+    public void checkThatApplicationServicesAreRunning() {
+        assertTrue( this.appService.servicesAreRunning() );
+    }
+}
+```
+
+This might not be working as expected. Why so?
```

Comment:
Let's create a full example that does work and describe all the parts needed to prevent confusion for the readers?